Fix unparser invalid sql for query with order

y-f-u opened the pull request:
## Which issue does this PR close?

This PR addresses two issues:
- missing order/limit/distinct on derived table
    Before the fix, query like `select a from (select a, b from table order by b desc) tmp order by a` will only keep the `order by b desc` at a wrong nesting level after unparsing.

- order by columns not on select
    Before the fix, query like `select a from table order by b` will be unparsed into `select table.a from (select table.a, table.b from table) order by table.b` which is invalid as the derived table misses alias.


## Rationale for this change

To prevent unparser generating invalid sql for above two common scenarios

## What changes are included in this PR?

- relation derive can happen for projection, limit, order, distinct as they can be top-level logical plan node.
- rewrite the logicalplan when it has a pattern which parsed from sort columns are not in select. See code comment for more details

## Are these changes tested?

Yes

## Are there any user-facing changes?

No

goldmedal commented on datafusion/sql/src/unparser/rewrite.rs:
```diff
@@ -99,3 +99,61 @@ fn rewrite_sort_expr_for_union(exprs: Vec<Expr>) -> Result<Vec<Expr>> {
+) -> Option<LogicalPlan> {
+    let mut collects = vec![];
+
+    for expr in p.expr.clone() {
+        collects.push(expr.clone());
+    }
```
```suggestion
    p.expr.iter().cloned().for_each(|expr| {
        collects.push(expr);
    });
```
I think this way could save one time to clone all expressions.

alamb replied:
In fact it looks like `p.expr` is a `Vec` already, so we simply deep-copy the vector like this

```rust
    let mut collects = p.expr.clone()
```

Since the code returns early for `Sort` and `Projection` it would be better to do the clone after those checks, otherwise the Exprs will be copied even if they are just then thrown away

goldmedal replied:
Indeed. It's better. Thanks.

y-f-u replied:
Thanks

goldmedal reviewed (approved):
Thanks @y-f-u. LGTM. Just one minor comment. I also tried other cases, and they work well.

alamb reviewed (approved):
Thank you @y-f-u  and @goldmedal  for the review

I think it would be good to implement some of the suggestions in this PR from @goldmedal but I think we could also do it as a follow on PR too.

y-f-u commented:
Thanks @alamb and @goldmedal . I've addressed all the comments, also fixed one missing case that the projected column is an alias. Now all works well :)

y-f-u commented on datafusion/sql/tests/cases/plan_to_sql.rs:
```diff
@@ -240,6 +240,63 @@ fn roundtrip_statement_with_dialect() -> Result<()> {
             unparser_dialect: Box::new(UnparserDefaultDialect {}),
         },
+        TestStatementWithDialect {
+            sql: "SELECT j1_string from j1 order by j1_id",
```
added to the test suite

y-f-u commented on datafusion/sql/tests/cases/plan_to_sql.rs:
```diff
@@ -240,6 +240,63 @@ fn roundtrip_statement_with_dialect() -> Result<()> {
+            parser_dialect: Box::new(GenericDialect {}),
+            unparser_dialect: Box::new(UnparserDefaultDialect {}),
+        },
+        // Test query with derived tables that put distinct,sort,limit on the wrong level
```
added to the test suite

alamb commented:
https://github.com/apache/datafusion/actions/runs/10001535587/job/27645407724?pr=11527

Doesn't seem related to the changes in this PR. I'll restart the check and if it passes let's file a ticket about the intermittent failure

```
failures:

---- fuzz_cases::join_fuzz::test_anti_join_1k_filtered stdout ----
thread 'fuzz_cases::join_fuzz::test_anti_join_1k_filtered' panicked at datafusion/core/tests/fuzz_cases/join_fuzz.rs:537:17:
assertion `left == right` failed: HashJoinExec and SortMergeJoinExec produced different row counts, batch_size: 1
  left: 952
 right: 953
stack backtrace:
```

alamb commented:
Filed https://github.com/apache/datafusion/issues/11555

alamb commented:
Oh no! It has another conflict to resolve. 

I can't push commits to the spiceai repo or I would fix it myself. 

<img width="1093" alt="Screenshot 2024-07-19 at 5 48 42 PM" src="https://github.com/user-attachments/assets/e6c23d17-44c1-4b06-be7a-3b1ef7c40866">

y-f-u commented:
I will look at it for a fix.

alamb commented:
Thanks @y-f-u
